### ISSUE TITLE
Why is `println` missing?

### ISSUE THREAD

**Issue by jiri** [fmtlib/fmt#428]
`fmt::println` seems like a natural addition to the API. Is there a reason why it's missing?

**Comment by foonathan** [comment]
What's wrong with `fmt::print("...\n")`?

**Comment by vitaut** [comment]
I think that `println` is of limited use because you can always specify the newline explicitly as @foonathan showed. But fmt provides sufficient infrastructure to easily and efficiently build this and similar methods without making them the part of the core library. In particular println can be written in five lines of code:

```c++
template <typename... Args>
void print(fmt::CStringRef format, const Args&... args) {
  fmt::print(format, args...);
  std::putc('\n', stdout);
}
```

**Comment by jiri** [comment]
Sure, and you could always use `stdout` explicitly, but it's the default file for convenience. One could as well argue "What's wrong with `fmt::print(stdout, "...\n")`.

Having to include an explicit newline everywhere is noisy and prone to forgetting them. There's a reason why many languages include `println` in their core libraries ;) Moreover, I thought that the point of a formatting library is eliminating the reason for all those non-newline terminated prints, since I could just format them into the strings, but nevertheless, you seem to be pretty set on not including it, so feel free to keep this closed.

Cheers.

**Comment by vitaut** [comment]
The difference from `fmt::print` overload taking `stdout` is that the latter saves a bit of typing

```c++
fmt::print("foo");
fmt::print(stdout, "foo");
```
and directly corresponds to `std::printf`. But you are right, it's somewhat redundant too.

**Comment by jiri** [comment]
It's true that `println` would save exactly zero characters. What's the approach to printing a foreign string? E.g., is something like

```cpp
void foo(std::string s) {
    fmt::print(s);
}
```

considered bad form or bad api usage? `println(s)` can make the intention clearer than `print("{}\n", s)` or god forbid `print(s + "\n")`.

**Comment by vitaut** [comment]
> is something like ... considered bad form or bad api usage? 

Yes. As with `printf`, it's better to write

```c++
void foo(std::string s) {
    fmt::print("{}", s);
}
```
Otherwise `s` will be parsed as a format string and will need to be escaped.

**Comment by slonm** [comment]
>What's wrong with `fmt::print("...\n")`?

For Windows it's: `fmt::print("...\r\n")`
Howto to hide this difference?
I'm vote for API extension

**Comment by vitaut** [comment]
You can use `fmt::print("...\n")` on Windows as well.

**Comment by foonathan** [comment]
The underlying streams will translate newlines, as long as they are not binary streams. stdout by default is not, so it will properly translate newlines.

**Comment by micfan** [comment]
Ummmm... I like `fmt::println()`

**Comment by slonm** [comment]
>What's wrong with fmt::print("...\n")?
It does not flush stream like it do std::endl

**Comment by eliaskosunen** [comment]
Although not mandated by the C standard, `stdout` is line-buffered by default on many implementations. This means, that a call to `fflush` is not necessary after `printf("...\n")` or `puts("...")`, and thus, `fmt::print("...\n")` would indeed flush the stream.

The same applies for `std::cout`, which _is_ mandated to be line-buffered, so writing a `'\n'` would flush the stream no matter what.

**Comment by slonm** [comment]
`stdout` is line buffered only if a stream refers to a terminal. As soon as you redirect it to a pipe or to a file, it will be block buffered.
Verified on Windows

**Comment by vitaut** [comment]
> It does not flush stream like it do std::endl

That's a feature. Even with hypothetical `println` you wouldn't want to always flush.

**Comment by egor10257** [comment]
Still why don't have flushing println()? 
It doesn't break anything, and if it is really unneeded nobody will use it (and vise versa).

**Comment by slonm** [comment]
>That's a feature. Even with hypothetical println you wouldn't want to always flush.

`fmt::print("...\n")` could be used if flush is not need.

**Comment by cup** [comment]
I agree that `println` should be implemented as well. Many other programming
languages offer such a method. For example C:

~~~c
#include <stdio.h>
int main() {
   puts("abcde");
}
~~~

C#:

~~~c
using System;
class Program {
   static void Main() {
      Console.WriteLine("abcde");
   }
}
~~~

Go
-----

~~~go
package main
import "fmt"
func main() {
   fmt.Println("abcde")
}
~~~

Rust:

~~~rs
fn main() {
   println!("abcde");
}
~~~

Nim
-----

~~~nim
echo "abcde"
~~~

Python:

~~~py
print('abcde')
~~~

**Comment by cup** [comment]
https://github.com/c42f/tinyformat/issues/57

**Comment by narodnik** [comment]
I honestly was shocked this isn't the default behaviour of fmt::print()

Usually if you don't want a newline, you can be explicit about this.

**Comment by theIDinside** [comment]
> `stdout` is line buffered only if a stream refers to a terminal. As soon as you redirect it to a pipe or to a file, it will be block buffered.
> Verified on Windows

Also verified on Linux Mint. When I do this for example in a Qt application, i **have** to put std::endl to be sure that it outputs immediately, otherwise it will arbitrarily show up.

**Comment by elvisdukaj** [comment]
Yeah it's annoying to explicitly flush the buffer...

**Comment by jiri** [comment]
I'd also like to point out that absence of such a basic ergonomic feature in the presence of features like the [literal based API](https://fmt.dev/latest/api.html#literal-based-api) is kinda funny 😄

**Comment by yesudeep** [comment]
The more common use case is println and it is so trivial it should be baked in. typing \n everywhere adds too much noise.

**Comment by hellow554** [comment]
I would speak against an implicit newline within `print`, because that's disturbing. But an extra `println` function wouldn't hurt. As you said, that can be easily done, so what's the reason not to do it? 
After all, it's all about ergonomics in a library @vitaut .

**Comment by jm4R** [comment]
What's the proper way to flush the buffer in current {fmt}? `std::fflush` from cstdio?

**Comment by hellow554** [comment]
@jm4R why do you hijack this thread instead of opening a new issue/question?

**Comment by Vox-Ben** [comment]
I've just been banging my head off this, I have a legacy logging system that I'm converting to use fmt, and I can't find any way to insert a newline at the end of log statements automatically. The problem is that @vitaut 's five-line println above (which is what I'd tried before finding this) doesn't work where it's being compiled into a library for use by other code. The compiler creates all the function signatures used by my logging code itself, and then as soon as some other piece of code imports the library and tries to pass something to the logger with a combination of types it didn't see during compilation of the logging library, the linker barfs saying it's got an undefined symbol.

So yeah. Absent a println it seems I have a choice between requiring all users of the library to start putting their newlines in themselves, or not porting it to fmt.

**Comment by realnc** [comment]
> I have a legacy logging system that I'm converting to use fmt, and I can't find any way to insert a newline at the end of log statements automatically.

Just write your logging functions using perfect forwarding like this:

```c++
template <typename... Args>
void logWarn(fmt::format_string<Args...>&& fmt_str, Args&&... args)
{
    fmt::print(stderr, "warning: {}\n", fmt::format(
        std::forward<fmt::format_string<Args...>>(fmt_str), std::forward<Args>(args)...));
}

template <typename... Args>
void logInfo(fmt::format_string<Args...>&& fmt_str, Args&&... args)
{
    fmt::print("info: {}\n", fmt::format(
        std::forward<fmt::format_string<Args...>>(fmt_str), std::forward<Args>(args)...));
}
```

C++20 automatic compile-time argument checking still works fine with these, and so does manual compile-time checking with `FMT_STRING`.

**Comment by technic** [comment]
I think println should be added. In any case many code-bases using this fmt library would write their own println one-liner. So why not to add it here? I've read this thread and didn't find any reasonable argument against it.

**Comment by jiri** [comment]
Given that this discussion has been ongoing for **6 years** now and that the developers behind fmtlib have shown zero interest in listening, I'd suggest any future commenters to just move onto other libraries with more active development.

I suggest @vitaut locks this issue, as any attempt to further convince you on this issue just adds noise in the repository activity at this point.

**Comment by vitaut** [comment]
> I've read this thread and didn't find any reasonable argument against it.

To add a feature there should be strong arguments for it not against not adding it. And so far the arguments have been somewhat lacking and some even became invalid with compile-time checks.

@jiri, no plans to lock this issue but at least please stop adding noise yourself.

**Comment by jiri** [comment]
@vitaut Please refrain from mentioning me again, I don't wish to continue this discussion with you. Thank you!

**Comment by technic** [comment]
> > I've read this thread and didn't find any reasonable argument against it.
> 
> To add a feature there should be strong arguments for it not against not adding it. And so far the arguments have been somewhat lacking and some even became invalid with compile-time checks.

I completely disagree to this approach of feature management.

**Comment by tryfinally** [comment]
1. its all about readability and explicit intention.
   println : I want to output sequence of chars ending with a new line,
2. commonality and principle of least astonishment, especially for us polyglots. ( C# , Java. to name a few)

**Comment by OnurKader** [comment]
Would it be reconsidered if implementers add `println` with [P2093](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2093r14.html)?

**Comment by emailstorbala** [comment]
Seems C++23 has included the `println()`.

https://en.cppreference.com/w/cpp/io/println

**Comment by hellow554** [comment]
@emailstorbala fmt also now has `println`
This issue is finally implemented

**Comment by emailstorbala** [comment]
> @emailstorbala fmt also now has `println` This issue is finally implemented

oh thank you @hellow554 ! I am using `Debian/trixie` with `libfmt9`  (9.1.0+ds1-2). Is this available in this version? Can you kindly point me to its documentation as I am unable to find it.

**Comment by tupaschoal** [comment]
> oh thank you @hellow554 ! I am using `Debian/trixie` with `libfmt9` (9.1.0+ds1-2). Is this available in this version? Can you kindly point me to its documentation as I am unable to find it.

As far as I can see on `git blame`, it was implemented for fmt 10.0.0, but not documented.
